Modules/_posixsubprocess.c

```diff
@@ -369,16 +392,36 @@ _close_open_fds_maybe_unsafe(long start_fd, PyObject* py_fds_to_keep)
         }
         if (errno) {
             /* readdir error, revert behavior. Highly Unlikely. */
-            _close_fds_by_brute_force(start_fd, py_fds_to_keep);
+            _close_range_except(start_fd, -1, py_fds_to_keep, _unsafe_closer);
         }
         closedir(proc_fd_dir);
     }
 }
 
-#define _close_open_fds _close_open_fds_maybe_unsafe
+#define _close_open_fds_fallback _close_open_fds_maybe_unsafe
 
 #endif  /* else NOT (defined(__linux__) && defined(HAVE_SYS_SYSCALL_H)) */
 
+#ifdef HAVE_CLOSE_RANGE
+static int
+_close_range_closer(int first, int last)
+{
+    return close_range(first, last, 0);
```

Comment:
What does glibc 2.34 close_range() actually do when the running kernel does not support the close_range system call?  does it return -1 with errno=ENOSYS [desirable] or does it implement its own potentially poor unsafe version of close_range?  The [Linux/glibc documentation](https://man7.org/linux/man-pages/man2/close_range.2.html) fails to be explicit about this.  Add a comment describing the findings to the code here for future reference. If it implements its own instead of just surfacing the syscall failure, we'd probably be better off making the syscall directly as we don't want to depend on such a glibc provided fallback when we've got our own tried and true tested efficient one.